manticore/core/smtlib/expression.py

```diff
@@ -512,8 +564,10 @@ def signed_value(self):
             return self._value
 
 
-class BitVecOperation(BitVec):
-    __slots__ = ["_operands"]
+class BitVecOperation(BitVec, abstract=True):
+    """ An operation that results in a BitVec """
+
+    __xslots__: Tuple[str, ...] = ("_operands",)
 
     def __init__(self, size, *operands, **kwargs):
```

Comment:
So, everywhere that you create a `BitVecOperation` or `BoolOperation`, you pass in `operands` as a tuple without unpacking it. However, the function signature doesn't change here. So now, by my estimation, `self._operands` becomes a tuple nested two stages deep. Is that intentional? It seems weird to me, but maybe it does something I'm not catching.

Reply:
Good thought. I see what you mean. This was pretty confusing for me when working on it tbh, so I'm glad we can talk about it more 😄 .

I'm going to think out loud a little bit to make sure we're on the same page though:

> everywhere that you create a `BitVecOperation` or `BoolOperation`, you pass in operands as a tuple without unpacking it.

For each subclass of `BitVecOperation` to correctly call its `__init__`, we need to do the following 
```python
super().__init__(size, operand1, operand2, ..., operandn, **kwargs)
```
which would pack `operand1, operand2` into `operands` as a tuple without any nesting. 

On the other hand, we shouldn't be doing this (which is what you're calling to attention, I think):
```python
super().__init__(size, tuple(operand1, operand2, ....), **kwargs)
```
right? This would double-pack the tuple if we left the `*operands` parameter as is in `BitVecOperation`s  `__init__`.

It's a similar deal with `BoolOperation`, just without the first `size` argument.

So now, we should check all subclasses of `BitVecOperation` and `BoolOperation` to see how we call the `super().__init__(...)`. 

... After some searching, I'm unable to manually find where we are packing the operands before passing to `__init__`. In that case, I'm going to add an `assert all(not isinstance(op, tuple) for op in operands)` statement to both `BitVecOperation` and `BoolOperation` `__init__` method and run the tests, which should catch any double tuple packing that happens.

All tests seem to pass on my machine still, so unless I'm still missing something, this PR doesn't do any tuple nesting. Would you like me to commit/add those assertion statements?

Reply:
The next PR https://github.com/trailofbits/manticore/pull/2395 _does_ take the `operands` as a pre-packed tuple and is keyword-only. That involved a lot more to refactor

Reply:
ping @ehennenfent for further discussion

Reply:
> After some searching, I'm unable to manually find where we are packing the operands before passing to __init__

You're right, I was looking at the various places where we construct `BitVecConcat`, which explicitly asks for a tuple in its type signature, which it then unpacks before passing up the inheritance tree.